src/test/java/com/google/devtools/build/lib/skyframe/RecursiveFilesystemTraversalFunctionTest.java

```diff
@@ -674,6 +684,22 @@ private void assertTraversalOfDirectory(Artifact directoryArtifact) throws Excep
     assertThat(v4).isEqualTo(v3);
     assertTraversalRootHashesAreEqual(v3, v4);
     assertThat(progressReceiver.invalidations).doesNotContain(traversalRoot);
+
+    // Add a new empty subdirectory to the directory and see that the value is rebuilt, but results
+    // in the same hash.
+    if (!directoryArtifact.isTreeArtifact()) {
+      childOf(directoryArtifact, "empty_dir").asPath().createDirectory();
+      if (directoryArtifact.isSourceArtifact()) {
+        invalidateDirectory(directoryArtifact);
+      } else {
+        invalidateOutputArtifact(directoryArtifact);
+      }
+      RecursiveFilesystemTraversalValue v5 =
+          traverseAndAssertFiles(traversalRoot, expected1, expected2, expected3);
+      assertThat(v5).isNotEqualTo(v4);
```

Comment:
I'm pretty confused by this test: I would not have expected `v4` and `v5` to be different here as the `TraversalRequest` in this function does not set `emitEmptyDirectoryNodes` to `true`.

In fact, `v5` and `v4` contain identical root and paths information, but don't satisfy referential equality - do you have an idea why?

Reply:
This is because of [`NestedSet` uses `Object[]::equals`](https://github.com/bazelbuild/bazel/blob/6b2915b5db8052a37bea84acb9ba94d3ba419230/src/main/java/com/google/devtools/build/lib/collect/nestedset/NestedSet.java#L847) which is really a reference equality. I see that in here, the elements are the same in those sets are equal, but the sets themselves are different objects.

To make matters more funny, in this case:

```
v4.getTransitiveFiles().equals(v5.getTransitiveFiles()) -> false
v4.getTransitiveFiles().shallowEquals(v5.getTransitiveFiles()) -> true
```

Realistically, you can assert the members of that:

```
assertThat(v5.getResolvedRoot()).isEqualTo(v4.getResolvedRoot());
assertThat(v5.getTransitiveFiles().toList()).containsExactlyElementsIn(v4.getTransitiveFiles().toList());
```

The only case of equals in this test is really when `v4` is the same instance as `v3`, namely, you could make the check in line 684 `assertThat(v4).isSameInstanceAs(v3)`. That is because we don't compute the traversal in this case.

Also, this test (existing code, not your fault!) has 2 problems--it fuses multitude of test cases together and manually parameterizes the test case (could use `@TestParameter`). You can consider adding a separate test just for the new thing not to make things worse. Having said that, this is not good already, hence you can leave what you have--I cannot reasonably expect you to take on our tech debt. If you are interested in having some better parameterization for this etc, please let me know, I will try to refactor that thing then. Again, feel free to punt/ignore that!